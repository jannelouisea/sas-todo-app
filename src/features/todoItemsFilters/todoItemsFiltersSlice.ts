import { createSlice, PayloadAction } from '@reduxjs/toolkit'

import { Sort } from 'src/static/enums'
import { ITodoItemsFilters } from 'src/interfaces';

const initialState: ITodoItemsFilters = {
    searchTerm: '',
    showNewer: true,
    showOlder: true,
    showCompleted: true,
    sortBy: Sort.Incmpl1st,
};

const todoItemsFiltersSlice = createSlice({
    name: 'todoItemsFilters',
    initialState,
    reducers: {
        searchTermChanged(state, action: PayloadAction<string>) {
            return { ...state, searchTerm: action.payload }
        },
        showNewerToggled(state) {
            return { ...state, showNewer: !state.showNewer }
        },
        showOlderToggled(state) {
            return { ...state, showOlder: !state.showOlder }
        },
        showCompletedToggled(state) {
            return { ...state, showCompleted: !state.showCompleted }
        },
        sortByChanged(state, action: PayloadAction<Sort>) {
            return { ...state, sortBy: action.payload }
        }
    }
});

export const {
    searchTermChanged,
    showNewerToggled,
    showOlderToggled,
    showCompletedToggled,
    sortByChanged,
} = todoItemsFiltersSlice.actions;

export default todoItemsFiltersSlice.reducer;