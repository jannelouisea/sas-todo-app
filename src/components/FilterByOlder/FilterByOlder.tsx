import {
    FilterByButton
} from 'src/components'
import HourglassBottomIcon from '@mui/icons-material/HourglassBottom';
import {
    MUIColor,
    MUIFontSize
} from 'src/static/enums'

import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { showOlderToggled } from 'src/features/todoItemsFilters/todoItemsFiltersSlice';
import PropTypes from 'prop-types';

type Props = {
    className: string
}

function FilterByOlder({ className }: Props) {
    const showOlder = useAppSelector(state => state.todoItemsFilters.showOlder);
    const dispatch = useAppDispatch();

    const onToggleFilter = () => {
        dispatch(showOlderToggled());
    }

    return <FilterByButton
        className={className}
        label='Older'
        color={MUIColor.Warning}
        startIcon={<HourglassBottomIcon fontSize={MUIFontSize.Small} />}
        filterEnabled={showOlder}
        onClick={onToggleFilter}
        ariaLabel='Filter by older'
    />;
}

FilterByOlder.propTypes = {
    className: PropTypes.string,
}

FilterByOlder.defaultProps = {
    className: ''
}

export default FilterByOlder;