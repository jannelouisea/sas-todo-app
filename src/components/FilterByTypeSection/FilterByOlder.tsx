import {
    FilterByButton
} from 'src/components'
import HourglassBottomIcon from '@mui/icons-material/HourglassTop';

import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { showOlderToggled } from 'src/features/todoItemsFilters/todoItemsFiltersSlice';
import PropTypes from 'prop-types';

type Props = {
    className: string
}

function FilterByOlder({ className }: Props) {
    const showOlder = useAppSelector(state => state.todoItemsFilters.showOlder);
    const dispatch = useAppDispatch();

    const onToggleFilter = () => {
        dispatch(showOlderToggled());
    }

    return <FilterByButton
        className={className}
        label='Older'
        color='warning'
        startIcon={<HourglassBottomIcon fontSize='small' />}
        filterEnabled={showOlder}
        onClick={onToggleFilter}
    />;
}

FilterByOlder.propTypes = {
    className: PropTypes.string,
}

FilterByOlder.defaultProps = {
    className: ''
}

export default FilterByOlder;