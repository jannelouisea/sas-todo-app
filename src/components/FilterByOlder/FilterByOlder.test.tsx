import FilterByOlder from './FilterByOlder';
import { Provider } from 'react-redux';
import {
    IStoreType,
    createStore
} from 'src/app/store';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import * as todoItemsFiltersActions from 'src/features/todoItemsFilters/todoItemsFiltersSlice';

describe('FilterByOlder', () => {
    let store: IStoreType;

    beforeEach(() => {
        store = createStore();
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <FilterByOlder className='foo' />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('toggles older filter on click', () => {
        const todoItemOlderSpy = jest.spyOn(todoItemsFiltersActions, 'showOlderToggled');

        render(
            <Provider store={store}>
                <FilterByOlder className='foo' />
            </Provider>
        );

        let button: HTMLInputElement = screen.getByLabelText('Filter by older');
        expect(button).toBeInTheDocument();
        fireEvent.click(button);

        expect(todoItemOlderSpy).toHaveBeenCalledTimes(1);
    })
});