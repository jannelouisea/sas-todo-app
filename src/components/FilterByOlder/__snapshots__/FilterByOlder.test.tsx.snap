// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterByOlder renders correctly 1`] = `
<div
  className="foo"
>
  <button
    aria-label="Filter by older"
    className="MuiButton-root MuiButton-variantSolid MuiButton-colorWarning MuiButton-sizeSm foo css-16bqlfy-JoyButton-root"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="MuiButton-startDecorator css-187ulk2-JoyButton-startDecorator"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
        data-testid="HourglassBottomIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m18 22-.01-6L14 12l3.99-4.01L18 2H6v6l4 4-4 3.99V22h12zM8 7.5V4h8v3.5l-4 4-4-4z"
        />
      </svg>
    </span>
    Older
  </button>
</div>
`;
